### ISSUE TITLE
Bottleneck of SRU

### ISSUE THREAD

**Issue by oiwio** [titu1994/keras-SRU#4]
I think the bottleneck of SRU may be not using CUDA level optimization, here is the official implementation of SRU:
[SRU](https://github.com/taolei87/sru/blob/master/cuda_functional.py)
But i don't know how to implement the CUDA operation in tensorflow or keras, do you know how to implement it? I want to use the SRU in my tensorflow program....
Thanks!

**Comment by titu1994** [comment]
No I don't know how to use CUDA ops in Keras. Keras is a high level abstraction over CUDA. It isnt possible.

**Comment by edmondja** [comment]
Do you think this is Cudnn optimized ? https://gist.github.com/tam17aki/1d5b3ad08201b7df48c7f25a169d0021

**Comment by oiwio** [comment]
@edmondja Thanks for reply. It's a tensorflow implement but not using CUDA level optimization.

> First, the matrix multiplications across all time steps can be batched, which can significantly improve the computation intensity and hence GPU utilization. Second, all element-wise operations of the sequence can be fused (compiled) into one kernel function and be parallelized across hidden dimensions. Without the fusion, operations such as addition + and sigmoid activation σ() would invoke a separate function call. This brings additional kernel launching latency and also adds data moving cost.

**Comment by edmondja** [comment]
Lool ok, so its shit compared to the work which has been made on keras-sru
